bin/recursive_gen/src/main.cpp

```diff
@@ -52,6 +53,58 @@
 #include <nil/blueprint/transpiler/public_input.hpp>
 #include <nil/blueprint/transpiler/recursive_verifier_generator.hpp>
 
+template<typename StreamType>
+std::optional<StreamType> open_file(const std::string& path, std::ios_base::openmode mode) {
+    StreamType file(path, mode);
+    if (!file.is_open()) {
+        std::cerr << "Unable to open file: " << path;
+        return std::nullopt;
+    }
+
+    return file;
+}
+
+std::optional<std::vector<std::uint8_t>> read_file_to_vector(const std::string& path) {
+
+    auto file = open_file<std::ifstream>(path, std::ios_base::in | std::ios::binary | std::ios::ate);
+    if (!file.has_value()) {
+        return std::nullopt;
+    }
+
+    std::ifstream& stream = file.value();
+    std::streamsize fsize = stream.tellg();
+    stream.seekg(0, std::ios::beg);
+    std::vector<std::uint8_t> v(static_cast<size_t>(fsize));
+
+    stream.read(reinterpret_cast<char*>(v.data()), fsize);
+
+    if (stream.fail()) {
+        std::cerr << "Error occurred during reading file " << path << std::endl;
+        return std::nullopt;
+    }
+
+    return v;
+}
+
+template<typename MarshallingType>
```

Comment:
same about extra copy, and what are you think about marshaling directly from file (i remember it not supported in crypto3, but possible?)

Reply:
Of course, it is possible. But I think, it'll be better to have separate issue for it))